### PR TITLE
Align DbConnexion and IDbConnexion + add Dapper async methods

### PR DESCRIPTION
**Pull request by Requette**
Add new features : 
QueryFirstAsync
QueryFirstOrDefaultAsync
QuerySingleAsync
QuerySingleOrDefaultAsync

Refactoring to unify DbConnexion and IDbConnexion management
Add multiple unit test

### CODE REVIEW

**Comment by Lonli-Lokli** [comment]
@Tewr 
is there any chance to review & merge it? 

@Requette 
Looking into new Mocks, will it be easy to add ExecuteScalarAsync & splitOn functionality into library?

**Comment by eeskildsen** [comment]
@Lonli-Lokli @Tewr @UnoSD Seconding the request to review and merge. This would be great to have.

**Comment by UnoSD** [comment]
As usual I have not had any time, but I will try and glance it now if it doesn't require too long. But thank you for the pull request.

File: Moq.Dapper.Test/DapperExecuteAsyncTest.cs [hunk 1]
```diff
@@ -12,15 +15,69 @@ public void ExecuteAsync()
...
-            connection.SetupDapperAsync(c => c.ExecuteAsync("", null, null, null, null))
+            connection.SetupDapperAsync(c => c.ExecuteAsync(It.IsAny<string>(), null, null, null, null))
                       .ReturnsAsync(1);
 
-            var result = connection.Object
-                                   .ExecuteAsync("")
-                                   .GetAwaiter()
-                                   .GetResult();
+            var actual = connection.Object
+                .ExecuteAsync("")
```
Comment (UnoSD) [inline Moq.Dapper.Test/DapperExecuteAsyncTest.cs]: Can you please keep the code style consistent? Indentation should match the dot from the previous line.

**Comment by Requette** [comment]
I modify the

File: Moq.Dapper.Test/DapperExecuteAsyncTest.cs (hunk 1 above)
Comment (Requette) [reply Moq.Dapper.Test/DapperExecuteAsyncTest.cs]: The code style is now consistent

File: Moq.Dapper.Test/DapperExecuteAsyncTest.cs [hunk 2]
```diff
@@ -12,15 +15,70 @@ public void ExecuteAsync()
         {
```
Comment (UnoSD) [inline Moq.Dapper.Test/DapperExecuteAsyncTest.cs]: This test seems to be broken

File: Moq.Dapper.Test/DapperExecuteAsyncTest.cs [hunk 3]
```diff
@@ -12,15 +14,70 @@ public void ExecuteAsync()
...
             var connection = new Mock<DbConnection>();
 
-            connection.SetupDapperAsync(c => c.ExecuteAsync("", null, null, null, null))
+            connection.SetupDapperAsync(c => c.ExecuteAsync(It.IsAny<string>(), null, null, null, null))
```
Comment (UnoSD) [inline Moq.Dapper.Test/DapperExecuteAsyncTest.cs]: This is better, but I'll revert it for not just to reduce the surface of changes together with result -> actual.

**Comment by UnoSD** [comment]
I spent some time looking at it, but there are a lot of non-functional changes (I.E. renaming of variables and moving stuff around etc) I've started rolling some back, but I ended up spending the time I had to remove those and not actually looking at the code. I'll try and get back to it this Saturday.

Before I look further, there is one test that is failing that stops me from merging it and building a new version.

Thanks.

**Comment by Lonli-Lokli** [comment]
@UnoSD is there any update? Thanks!

**Comment by UnoSD** [comment]
@Lonli-Lokli I managed to get some time to have a look at it, but unfortunately a test is broken so I can't merge it until it's fixed.

**Comment by Lonli-Lokli** [comment]
@UnoSD do you need help from @Requette  or community to fix it?

**Comment by UnoSD** [comment]
@Lonli-Lokli any help is appreciated, I haven't touched it in a while so it takes me even longer to figure out how to review PRs.

I put some effort in removing all non-functional changes from here (variable renames, reformatting, moving code), but in the end I realised tests were not passing.

I don't have time to fix it myself; I'll do the best I can to integrate changes from the community, but they need to be consistent with the repo's code style and, obviously, shouldn't break existing code.

**Comment by Lonli-Lokli** [comment]
@UnoSD I do not have rights for @Requette  repository, but failing test can be fixed with this code:
Moq.Dapper/DbCommandSetup.cs

before:
```
            mockResult(commandMock, () => result); 

            mock
                .Setup(m => m.CreateCommand())
                .Returns(commandMock.Object);
 
            return setupMock.Object;
```

after: 

```
            mockResult(commandMock, () => result); 

            mock
                .As<IDbConnection>()
                .Setup(m => m.CreateCommand())
                .Returns(commandMock.Object);
 
            return setupMock.Object;
```

**Comment by Requette** [comment]
Hello,

I will update the PR the weekend.

**Comment by UnoSD** [comment]
Thanks @Lonli-Lokli @Requette

**Comment by Requette** [comment]
I fix the TU and the property name, even if it isn't a functional change, I think I could only be better.

**Comment by UnoSD** [comment]
@Requette I agree it needs to be renamed, but I prefer not to have mixed commits. I'll make the change separately.

**Comment by Requette** [comment]
Ok no problem

**Comment by Lonli-Lokli** [comment]
@UnoSD as for me this PR is ok, just maybe one of you will update readme with sample usage as well?

**Comment by Requette** [comment]
I add some new method example (the list isn't exhaustive)

File: Moq.Dapper.Test/Moq.Dapper.Test.csproj [hunk 4]
```diff
@@ -39,24 +40,37 @@
...
+      <HintPath>..\packages\Castle.Core.4.4.0\lib\net45\Castle.Core.dll</HintPath>
     </Reference>
-    <Reference Include="Dapper, Version=1.50.0.0, Culture=neutral, processorArchitecture=MSIL">
-      <HintPath>..\packages\Dapper.1.50.7\lib\net451\Dapper.dll</HintPath>
+    <Reference Include="Dapper, Version=1.60.0.0, Culture=neutral, processorArchitecture=MSIL">
+      <HintPath>..\packages\Dapper.1.60.6\lib\net451\Dapper.dll</HintPath>
     </Reference>
-    <Reference Include="Moq, Version=4.7.1.0, Culture=neutral, PublicKeyToken=69f491c39445e920, processorArchitecture=MSIL">
-      <HintPath>..\packages\Moq.4.7.1\lib\net45\Moq.dll</HintPath>
+    <Reference Include="Microsoft.CSharp" />
```
Comment (UnoSD) [inline Moq.Dapper.Test/Moq.Dapper.Test.csproj]: Why?

File: Moq.Dapper.Test/Moq.Dapper.Test.csproj [hunk 5]
```diff
@@ -39,24 +40,37 @@
...
+    <Reference Include="Moq, Version=4.11.0.0, Culture=neutral, PublicKeyToken=69f491c39445e920, processorArchitecture=MSIL">
+      <HintPath>..\packages\Moq.4.11.0\lib\net45\Moq.dll</HintPath>
     </Reference>
-    <Reference Include="nunit.framework, Version=3.6.1.0, Culture=neutral, PublicKeyToken=2638cd05610744eb, processorArchitecture=MSIL">
-      <HintPath>..\packages\NUnit.3.6.1\lib\net45\nunit.framework.dll</HintPath>
+    <Reference Include="nunit.framework, Version=3.12.0.0, Culture=neutral, PublicKeyToken=2638cd05610744eb, processorArchitecture=MSIL">
+      <HintPath>..\packages\NUnit.3.12.0\lib\net45\nunit.framework.dll</HintPath>
     </Reference>
     <Reference Include="System" />
+    <Reference Include="System.Configuration" />
```
Comment (UnoSD) [inline Moq.Dapper.Test/Moq.Dapper.Test.csproj]: Why?

**Comment by UnoSD** [comment]
Sorry, I tried really hard but I can't merge this.

I would reconsider it if:

1) Changes are kept to the minimum, no "change1" + "change2" PRs
2) Refactoring is fine, but not in the same PR as other changes; classes moved, property renamed, packages updated, metadata changes cannot be in the same PR, I cannot spend my limited time reverting those

I am not using this library anymore therefore I am only doing this for the community in my limited spare time, if I have to take time to merge a PR I would like to see the smallest functional change + it's corresponding test so I don't have to waste days trying to understand what's going on and reverting other changes.

I tried to spend an hour today to merge it because it seems like something people would really like to see in the library, but I can't spend more time on this. Please consider removing ALL changes unrelated to the functionality added and split those into the smallest chunks possible in multiple PRs; in that case I would be happy to merge it, but, as it is, I am going to have to close it.

**Comment by Lonli-Lokli** [comment]
This PR has all functions mentioned in first post, no more and no less. I do not get why multiple goals MUST be in separate PRs.

**Comment by UnoSD** [comment]
I need to see the changes in isolation so it is as easy and fast as possible for me to merge it, I do not have time to review 20 files and also fix code style and all other random changes in the branch. I tried, but I ended up again spending time revering renamed properties and all the other stuff I mentioned above. I'd like to remind I do this for free so it should, at very least, be facilitated by the others. I appreciate the contributions, but I rather spend my weekends doing something else honestly.

**Pull request closed without merging**